ord_schema/interface/query.py

```diff
@@ -20,15 +47,252 @@
 from absl import logging
 import psycopg2
 from psycopg2 import sql
+from rdkit import Chem
 
 from ord_schema.proto import dataset_pb2
 from ord_schema.proto import reaction_pb2
 
-# pylint: disable=too-many-locals
 
+def fetch_results(cursor):
+    """Fetches query results.
+
+    Args:
+        cursor: psycopg.cursor instance.
+
+    Returns:
+        Dict mapping reaction IDs to serialized Reaction protos.
+    """
+    reactions = {}
+    for reaction_id, serialized in cursor:
+        reactions[reaction_id] = serialized
+    return reactions
+
+
+class ReactionQueryBase:
+    """Base class for reaction-based queries."""
+
+    def json(self):
+        """Returns a JSON representation of the query."""
+        raise NotImplementedError()
+
+    def run(self, cursor, limit=None):
+        """Runs the query.
+
+        Args:
+            cursor: psycopg.cursor instance.
+            limit: Integer maximum number of matches. If None (the default), no
+                limit is set.
+
+        Returns:
+            Dict mapping reaction IDs to serialized Reaction protos.
+        """
+        raise NotImplementedError()
+
+
+class ReactionIdQuery(ReactionQueryBase):
+    """Looks up reactions by ID."""
+
+    def __init__(self, reaction_ids):
+        """Initializes the query.
+
+        Args:
+            reaction_ids: List of reaction IDs.
+        """
+        self._reaction_ids = reaction_ids
+
+    def json(self):
+        """Returns a JSON representation of the query."""
+        return json.dumps({'reactionIds': self._reaction_ids})
+
+    def run(self, cursor, limit=None):
+        """Runs the query.
+
+        Args:
+            cursor: psycopg.cursor instance.
+            limit: Not used; present for compatibility.
+
+        Returns:
+            Dict mapping reaction IDs to serialized Reaction protos.
+        """
+        del limit  # Unused.
+        query = sql.SQL("""
+            SELECT DISTINCT reaction_id, serialized 
+            FROM reactions 
+            WHERE reaction_id = ANY (%s)""")
+        args = [self._reaction_ids]
+        logging.info('Running SQL command:%s',
+                     cursor.mogrify(query, args).decode())
+        cursor.execute(query, args)
```

Comment:
Does this do the right thing in case of multiple reaction IDs? I don't know psycopg, but I would have expected something to say how to convert the list to a string, e.g. a delimiter.

Reply:
Yes, note that `self._reaction_ids` is a list, so `args` is a list of lists. See also `test_reaction_id_query` in `query_longtest.py`.